File: sdk/c/src/normalize.c
/**
 * 邮件数据标准化模块
 */

#include "tempmail_internal.h"
#include <time.h>

/* 将 received_at 等字段（字符串或毫秒/秒时间戳）格式化为 RFC3339 UTC */
static char *tm_normalize_date_json(const cJSON *raw) {
    const char *k1[] = {"received_at", "receivedAt", "created_at", "createdAt", "date"};
    for (int i = 0; i < 5; i++) {
        const cJSON *item = cJSON_GetObjectItemCaseSensitive(raw, k1[i]);
        if (!item) continue;
        if (cJSON_IsString(item) && item->valuestring && item->valuestring[0] != '\0') {
            return tm_strdup(item->valuestring);
        }
        if (cJSON_IsNumber(item) && item->valuedouble > 0) {
            double v = item->valuedouble;
            time_t sec = (v > 1e12) ? (time_t)(v / 1000.0 + 0.5) : (time_t)(v + 0.5);
            char buf[48];
            struct tm tmu;
#ifdef _WIN32
            if (gmtime_s(&tmu, &sec) != 0) memset(&tmu, 0, sizeof(tmu));
#else
            gmtime_r(&sec, &tmu);
#endif
            strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%SZ", &tmu);
            return tm_strdup(buf);
        }
    }
    const cJSON *ts = cJSON_GetObjectItemCaseSensitive(raw, "timestamp");
    if (cJSON_IsNumber(ts) && ts->valuedouble > 0) {
        double v = ts->valuedouble;
        time_t sec = (v < 1e12) ? (time_t)(v + 0.5) : (time_t)(v / 1000.0 + 0.5);
        char buf[48];
        struct tm tmu;
#ifdef _WIN32
        if (gmtime_s(&tmu, &sec) != 0) memset(&tmu, 0, sizeof(tmu));
#else
        gmtime_r(&sec, &tmu);
#endif
        strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%SZ", &tmu);
        return tm_strdup(buf);
    }
    const cJSON *ed = cJSON_GetObjectItemCaseSensitive(raw, "e_date");
    if (cJSON_IsNumber(ed) && ed->valuedouble > 0) {
        time_t sec = (time_t)(ed->valuedouble / 1000.0 + 0.5);
        char buf[48];
        struct tm tmu;
#ifdef _WIN32
        if (gmtime_s(&tmu, &sec) != 0) memset(&tmu, 0, sizeof(tmu));
#else
        gmtime_r(&sec, &tmu);
#endif
        strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%SZ", &tmu);
        return tm_strdup(buf);
    }
    return tm_strdup("");
}

tm_email_t tm_normalize_email(const cJSON *raw, const char *recipient) {
    tm_email_t email;
    memset(&email, 0, sizeof(email));

    /* ID */
    const char *id_keys[] = {"id", "eid", "_id", "mailboxId", "messageId", "mail_id"};
    email.id = tm_json_get_str(raw, id_keys, 6);

    /* From */
    const char *from_keys[] = {"from_addr", "from_address", "fromAddress", "mail_sender", "sender", "address_from", "from_email", "from", "messageFrom"};
    email.from_addr = tm_json_get_str(raw, from_keys, 9);

    /* To */
    const char *to_keys[] = {"to", "to_address", "toAddress", "name_to", "email_address", "address"};
    email.to = tm_json_get_str(raw, to_keys, 6);
    if (!email.to || email.to[0] == '\0') {
        free(email.to);
        email.to = tm_strdup(recipient);
    }

    /* Subject */
    const char *subj_keys[] = {"subject", "e_subject", "mail_title"};
    email.subject = tm_json_get_str(raw, subj_keys, 3);

    /* Text */
    const char *text_keys[] = {"text", "text_body", "preview_text", "mail_body_text", "body", "content", "body_text", "text_content", "description"};
    email.text = tm_json_get_str(raw, text_keys, 9);

    /* HTML */
    const char *html_keys[] = {"html", "html_body", "html_content", "body_html", "mail_body_html"};
    email.html = tm_json_get_str(raw, html_keys, 5);

    /*
     * 修正 text/html 错配：部分渠道将 HTML 放在 body/text 字段中
     * 检测 text 是否包含 HTML 标签，如果是则移动到 html 字段
     */
    if (email.text && email.text[0] != '\0' &&
        (!email.html || email.html[0] == '\0')) {
        const char *lower = email.text;
        int is_html = 0;
        if (strstr(lower, "<!DOCTYPE") || strstr(lower, "<!doctype") ||
            strstr(lower, "<html") || strstr(lower, "<HTML") ||
            strstr(lower, "<body") || strstr(lower, "<BODY") ||
            (strstr(lower, "<div") && strstr(lower, "</div>")) ||
            (strstr(lower, "<table") && strstr(lower, "</table>")) ||
            (strstr(lower, "<p") && strstr(lower, "</p>"))) {
            is_html = 1;
        }
        if (is_html) {
            free(email.html);
            email.html = email.text;
            email.text = tm_strdup("");
        }
    }

    /* Date（含数字型毫秒时间戳，如 ta-easy received_at） */
    email.date = tm_normalize_date_json(raw);

    /* IsRead */
    email.is_read = false;
    const char *read_keys[] = {"seen", "read", "isRead", "is_read", "is_seen"};
    for (int i = 0; i < 4; i++) {
        const cJSON *item = cJSON_GetObjectItemCaseSensitive(raw, read_keys[i]);
        if (cJSON_IsBool(item)) {
            email.is_read = cJSON_IsTrue(item);
            break;
        }
        if (cJSON_IsNumber(item)) {
            email.is_read = item->valueint != 0;
            break;
        }
    }

    /* Attachments */
    email.attachments = NULL;
    email.attachment_count = 0;
    const cJSON *att_arr = cJSON_GetObjectItemCaseSensitive(raw, "attachments");
    if (cJSON_IsArray(att_arr)) {
        int count = cJSON_GetArraySize(att_arr);
        if (count > 0) {
            email.attachments = (tm_attachment_t*)calloc(count, sizeof(tm_attachment_t));
            if (email.attachments) {
                email.attachment_count = count;
                int idx = 0;
                const cJSON *a;
                cJSON_ArrayForEach(a, att_arr) {
                    const char *fn_keys[] = {"filename", "name"};
                    email.attachments[idx].filename = tm_json_get_str(a, fn_keys, 2);

                    const cJSON *sz = cJSON_GetObjectItemCaseSensitive(a, "size");
                    if (!sz) sz = cJSON_GetObjectItemCaseSensitive(a, "filesize");
                    email.attachments[idx].size = sz && cJSON_IsNumber(sz) ? (long)sz->valuedouble : -1;

                    const char *ct_keys[] = {"contentType", "content_type", "mimeType", "mime_type"};
                    email.attachments[idx].content_type = tm_json_get_str(a, ct_keys, 4);

                    const char *url_keys[] = {"url", "download_url", "downloadUrl"};
                    email.attachments[idx].url = tm_json_get_str(a, url_keys, 3);
                    idx++;
                }
            }
        }
    }

    return email;
}

File: sdk/php/src/Normalize.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail;

/**
 * 邮件数据标准化模块
 *
 * 不同渠道的 API 返回字段名各不相同，本类通过多候选字段策略将其统一映射为
 * 标准的 Email 结构，保证 SDK 输出一致性。
 */
final class Normalize
{
    /** @var string[] 正文纯文本字段候选（按优先级），提为常量避免每次调用重复分配数组 */
    private const TEXT_KEYS = ['text', 'text_body', 'preview_text', 'mail_body_text', 'body', 'content', 'body_text', 'text_content', 'description'];

    /** @var string[] 正文 HTML 字段候选 */
    private const HTML_KEYS = ['html', 'html_body', 'html_content', 'body_html', 'mail_body_html'];

    /** @var string[] 邮件 ID 字段候选 */
    private const ID_KEYS = ['id', 'eid', '_id', 'mailboxId', 'messageId', 'mail_id'];

    /** @var string[] 发件人字段候选 */
    private const FROM_KEYS = ['from_addr', 'from_address', 'fromAddress', 'mail_sender', 'sender', 'address_from', 'from_email', 'from', 'messageFrom'];

    /** @var string[] 收件人字段候选 */
    private const TO_KEYS = ['to', 'to_address', 'toAddress', 'name_to', 'email_address', 'address'];

    /** @var string[] 主题字段候选 */
    private const SUBJECT_KEYS = ['subject', 'e_subject', 'mail_title'];

    /**
     * 将各提供商返回的原始邮件数据标准化为统一的 Email 格式。
     *
     * @param array<mixed> $raw            原始邮件数据
     * @param string       $recipientEmail 收件人地址（无匹配字段时回退）
     */
    public static function email(array $raw, string $recipientEmail = ''): Email
    {
        $text = self::pick($raw, self::TEXT_KEYS);
        $html = self::pick($raw, self::HTML_KEYS);

        // 修正 text/html 错配
        if ($text !== '' && $html === '' && self::isHtml($text)) {
            $html = $text;
            $text = '';
        }
        if ($text === '' && $html !== '') {
            $text = self::htmlToText($html);
        }
        if ($text !== '' && $html === '') {
            $html = '<html><body><pre>' . htmlspecialchars($text, ENT_QUOTES) . '</pre></body></html>';
        }

        return new Email(
            id: self::pick($raw, self::ID_KEYS),
            fromAddr: self::pick($raw, self::FROM_KEYS),
            to: self::pick($raw, self::TO_KEYS) ?: $recipientEmail,
            subject: self::pick($raw, self::SUBJECT_KEYS),
            text: $text,
            html: $html,
            date: self::normalizeDate($raw),
            isRead: self::normalizeIsRead($raw),
            attachments: self::normalizeAttachments($raw['attachments'] ?? null),
        );
    }

    /**
     * 从数组中按优先级顺序提取字符串值。
     *
     * @param array<mixed> $raw
     * @param string[]     $keys
     */
    private static function pick(array $raw, array $keys): string
    {
        foreach ($keys as $key) {
            if (array_key_exists($key, $raw) && $raw[$key] !== null) {
                $val = $raw[$key];
                if (is_string($val)) {
                    return $val;
                }
                if (is_scalar($val)) {
                    return (string) $val;
                }
            }
        }
        return '';
    }

    /**
     * 检测内容是否为 HTML。
     */
    private static function isHtml(string $content): bool
    {
        $prefix = strtolower(ltrim(substr($content, 0, 200)));
        if (str_starts_with($prefix, '<!doctype html') || str_starts_with($prefix, '<html') || str_starts_with($prefix, '<body')) {
            return true;
        }
        $trimmed = strtolower(trim($content));
        if (str_contains($trimmed, '<div') && str_contains($trimmed, '</div>')) {
            return true;
        }
        if (str_contains($trimmed, '<table') && str_contains($trimmed, '</table>')) {
            return true;
        }
        if (str_contains($trimmed, '<p') && str_contains($trimmed, '</p>') && str_contains($trimmed, '<')) {
            return true;
        }
        return false;
    }

    /**
     * 将 HTML 粗略转换为纯文本。
     */
    private static function htmlToText(string $html): string
    {
        $s = preg_replace('/<(script|style)[^>]*>.*?<\/\1>/is', '', $html) ?? $html;
        $s = preg_replace('/<br\s*\/?>/i', "\n", $s) ?? $s;
        $s = preg_replace('/<\/(p|div|tr|li|h[1-6])>/i', "\n", $s) ?? $s;
        $s = strip_tags($s);
        $s = html_entity_decode($s, ENT_QUOTES | ENT_HTML5);
        $s = preg_replace('/[ \t]+/', ' ', $s) ?? $s;
        $s = preg_replace('/\n{3,}/', "\n\n", $s) ?? $s;
        return trim($s);
    }

    /**
     * 提取并统一日期格式为 ISO 8601，支持字符串日期与秒/毫秒时间戳。
     *
     * @param array<mixed> $raw
     */
    private static function normalizeDate(array $raw): string
    {
        foreach (['received_at', 'receivedAt', 'created_at', 'createdAt', 'date'] as $key) {
            $val = $raw[$key] ?? null;
            if ($val === null) {
                continue;
            }
            if (is_string($val) && $val !== '') {
                $ts = strtotime($val);
                if ($ts !== false) {
                    return date('c', $ts);
                }
                return $val;
            }
            if ((is_int($val) || is_float($val)) && $val > 0) {
                $seconds = $val > 1e12 ? (int) ($val / 1000) : (int) $val;
                return date('c', $seconds);
            }
        }

        foreach (['timestamp', 'e_date'] as $key) {
            $val = $raw[$key] ?? null;
            if ((is_int($val) || is_float($val)) && $val > 0) {
                if ($key === 'timestamp' && $val < 1e12) {
                    return date('c', (int) $val);
                }
                return date('c', (int) ($val / 1000));
            }
        }

        return '';
    }

    /**
     * 提取已读状态，支持 bool / int(0|1) / str('0'|'1')。
     *
     * @param array<mixed> $raw
     */
    private static function normalizeIsRead(array $raw): bool
    {
        foreach (['seen', 'read', 'isRead'] as $key) {
            $val = $raw[$key] ?? null;
            if (is_bool($val)) {
                return $val;
            }
        }
        foreach (['is_read', 'is_seen'] as $key) {
            $val = $raw[$key] ?? null;
            if ($val === null) {
                continue;
            }
            if (is_bool($val)) {
                return $val;
            }
            if (is_int($val) || is_float($val)) {
                return (int) $val !== 0;
            }
            if (is_string($val)) {
                return $val === '1';
            }
        }
        return false;
    }

    /**
     * 提取并标准化附件列表。
     *
     * @param mixed $attachments
     * @return EmailAttachment[]
     */
    private static function normalizeAttachments(mixed $attachments): array
    {
        if (!is_array($attachments)) {
            return [];
        }
        $result = [];
        foreach ($attachments as $a) {
            if (!is_array($a)) {
                continue;
            }
            $size = $a['size'] ?? $a['filesize'] ?? null;
            $result[] = new EmailAttachment(
                filename: (string) ($a['filename'] ?? $a['name'] ?? ''),
                size: is_numeric($size) ? (int) $size : null,
                contentType: $a['contentType'] ?? $a['content_type'] ?? $a['mimeType'] ?? $a['mime_type'] ?? null,
                url: $a['url'] ?? $a['download_url'] ?? $a['downloadUrl'] ?? null,
            );
        }
        return $result;
    }
}
